### PR TITLE
Vision conversion

### PR DESCRIPTION
**Pull request by bramstoeller**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -76,11 +77,11 @@ power_grid_model_io = ["*.yaml"]
 
 [tool.pytest.ini_options]
 testpaths = ["tests/unit"]
-addopts = ["--cov=src/power_grid_model_io", "--cov-report", "term", "--cov-fail-under=20"]
+addopts = ["--cov=src/power_grid_model_io", "--cov-report", "term", "--cov-report", "html:python_cov_html", "--cov-report", "xml:python_coverage.xml", "--cov-fail-under=20"]
```
Comment (bramstoeller) [inline pyproject.toml:80]: Do we really need the html covarage report?
Can we name the coverage report `coverage.xml`; this is a python only project.

**Review by bramstoeller — COMMENTED** [review]
In the .gitignore file: I think it's a bit messy to just ignore everything that could come up. Can we be a bit more selective?

File: pyproject.toml [hunk 2]
```diff
@@ -76,11 +77,11 @@ power_grid_model_io = ["*.yaml"]
...
 
 [tool.black]
 line-length = 120
-target-version = ["py39", "py310", "py311"]
+target-version = ["py39"]
```
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply pyproject.toml:84]: Done

File: pyproject.toml (hunk 1 above)
Comment (TonyXiang8787) [reply pyproject.toml]: xml is needed later for sonarcloud.

html is not per se needed. I personally find it nice to read the coverage report in html.

Comment (TonyXiang8787) [reply pyproject.toml]: removed python prefix.

**Comment by TonyXiang8787** [comment]
@bramstoeller  two general comments:

1. Did you check licenses of all the dependencies. Are they compatible with our MPL2 license?
2. I am not so sure about optional dependencies. Can we just include all the dependencies as default?

File: pyproject.toml [hunk 3]
```diff
@@ -40,7 +43,27 @@ dev = [
...
+    "openpyxl",
+    "pyyaml",
+    "typer[all]",
+]
+
+[project.urls]
+Home-page = "https://github.com/alliander-opensource/power-grid-model-io"
+
+[project.scripts]
+convert = "power_grid_model_io.cli.convert:app"
```
Comment (TonyXiang8787) [inline pyproject.toml:65]: the script command will be inject into the environment. I would call them with a prefix like `pgm_convert` or `pgm_valdiate`. Otherwise we may have a name conflict with other system command.

File: src/power_grid_model_io/utils/filters.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    "YN": WindingType.wye_n,
+    "D": WindingType.delta,
+    "Z": WindingType.wye,
+    "ZN": WindingType.wye_n,
+}
+
+T = TypeVar("T")
+
+
+def relative_no_load_current(i: float, p: float, s_nom: float, u_nom: float) -> float:
```
Comment (TonyXiang8787) [inline src/power_grid_model_io/utils/filters.py:31]: @bramstoeller  since the conversion performance will be essential (for pandapower). We need to have vectorized version of all these filter functions. The input/output should be numpy arrays.

File: src/power_grid_model_io/data_stores/base_data_store.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        self._log = structlog.get_logger(type(self).__name__)
+
+    @abstractmethod  # pragma: no cover
+    def load(self) -> T:
+        """
+        The method that loads the data from one or more sources and returns it in the specified format.
+        Note that the load() method does not recieve a reference to the data source(s); i.e. the data souce(s)
+        should be set in the constructor, or in a separate member method.
+        """
```
Comment (TonyXiang8787) [inline src/power_grid_model_io/data_stores/base_data_store.py:33]: an empty method should have a `pass` in the body.

File: src/power_grid_model_io/cli/convert.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+import structlog
+from power_grid_model.validation import errors_to_string, validate_input_data
+
+from power_grid_model_io.converters.gaia_excel_converter import GaiaExcelConverter
+from power_grid_model_io.converters.pgm_converter import PgmConverter
+from power_grid_model_io.converters.vision_excel_converter import VisionExcelConverter
+from power_grid_model_io.utils.modules import import_optional_module
+
+typer = import_optional_module("cli", "typer")
```
Comment (petersalemink95) [inline src/power_grid_model_io/cli/convert.py:22]: I'd prefer to specify the parameters here, e.g. `typer = import_optional_module(extra="cli", module="typer")`

File: src/power_grid_model_io/data_stores/base_data_store.py (hunk 5 above)
Comment (bramstoeller) [reply src/power_grid_model_io/data_stores/base_data_store.py:33]: Not necessarily; `pass` is just a placeholder to fix the syntax. In this case, the comment serves as a placeholder. I'm not entirely sure, but I guess the liner (pylint) would complain about an extra `pass`.

File: src/power_grid_model_io/cli/convert.py (hunk 6 above)
Comment (bramstoeller) [reply src/power_grid_model_io/cli/convert.py:22]: Yes, I agree. I'll fix it.

File: pyproject.toml (hunk 3 above)
Comment (bramstoeller) [reply pyproject.toml]: Yes, fair point. I'll fix it.

File: src/power_grid_model_io/utils/filters.py (hunk 4 above)
Comment (bramstoeller) [reply src/power_grid_model_io/utils/filters.py:31]: I agree, partially.

My rationale was: Create a way to define functions in an understandable way. These mapping functions are much easier to read and understand than vectorized function. Currently we didn't have any efficiency constraints, but indeed, they'll come up in the near future. Let's create new stories for it.

Peter and I do have some ideas about efficiency (for pandapower), let's discuss it in a refinement meeting.

**Comment by bramstoeller** [comment]
> @bramstoeller two general comments:
> 
> 1. Did you check licenses of all the dependencies. Are they compatible with our MPL2 license?
No, I didn't. I'll check them.
> 2. I am not so sure about optional dependencies. Can we just include all the dependencies as default?
It was a requirement from one of our users to "please make sure that the conversions library doesn't get super heavy".
We could discuss how to organize the extra requirements.

**Comment by TonyXiang8787** [comment]
@bramstoeller 
https://github.com/Alliander/python_scientific_computing/blob/master/Tabular_data_manipulation_in_python.ipynb

File: examples/pgm_conversion.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+from power_grid_model import PowerGridModel
+
+from power_grid_model_io.converters.pgm_converter import PgmConverter
+
+DATA_DIR = Path(__file__).parent / "data"
+
+if __name__ == "__main__":
+    structlog.configure(wrapper_class=structlog.make_filtering_bound_logger(logging.DEBUG))
+
+    input_file = DATA_DIR / "1os2msr" / "input.json"
```
Comment (petersalemink95) [inline examples/pgm_conversion.py:19]: I think we should rename these examples

File: examples/pgm_conversion.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    output_file = DATA_DIR / "1os2msr" / "sym_output.json"
```
Comment (petersalemink95) [inline examples/pgm_conversion.py:20]: This one as well

File: examples/pgm_conversion.py (hunk 7 above)
Comment (petersalemink95) [reply examples/pgm_conversion.py:19]: 1os2msr doesn't mean anything in English

File: .gitignore [hunk 9]
```diff
@@ -2,17 +2,146 @@
...
 
-# Generated files
-*.egg-info
-
-# temporary_files
+# Byte-compiled / optimized / DLL files
 __pycache__/
-.mypy_cache/
-.pytest_cache/
+*.py[cod]
```
Comment (petersalemink95) [inline .gitignore:7]: @TonyXiang8787 This comment counts for all things that have changed in the `.gitignore`.

Why did you add so many things to ignore? I think this is an overkill. We should only add things to the gitignore when you plan to use is and don't want it to be commited. I think it doesn't make sense to try to add all the things that you can think of to the gitignore.

File: docs/converters/tabular_converter.md [hunk 10]
```diff
@@ -0,0 +1,186 @@
+<!--
+SPDX-FileCopyrightText: 2022 Contributors to the Power Grid Model IO project <dynamic.grid.calculation@alliander.com>
+
+SPDX-License-Identifier: MPL-2.0
+-->
+# Tabular converter
+Tabular data can come from Excel files, a set of CSV files, GNF files, databases, pandas DataFrames, etc etc.
+The similarity between all tabular data is that it contains multiple `tabled`,
```
Comment (petersalemink95) [inline docs/converters/tabular_converter.md:8]: *typo: tables

File: docs/converters/tabular_converter.md [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+substitutions:
+  ".*Switch state":
+    "off": 0
+    "on": 1
+```
+
+## Grid
+For each `table`, the target PGM `component` is listed (e.g. Nodes: node, Cables: line).
+The for each PFM `column` the source column is supplied (e.g. u_rated: Unom, from_status: From.SwitchStatus).
```
Comment (petersalemink95) [inline docs/converters/tabular_converter.md:45]: *typo PGM

File: docs/converters/tabular_converter.md [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+but in many cases the mapping is a bit more complex.
+You can use the following `column` definitions: 
+
+  * Column name `str`
+    ```yaml
+    from_node: From.Number
+    ```
+  * First matching column name `str`
+    ```yaml
+    from_node: Inverter.Pnom | Inverter.Snom
```
Comment (petersalemink95) [inline docs/converters/tabular_converter.md:59]: `from_node` should be `p_specified`

File: examples/pgm_conversion.py (hunk 7 above)
Comment (bramstoeller) [reply examples/pgm_conversion.py]: Done, I've called it tiny-net.

File: docs/converters/tabular_converter.md [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+        table["Inverter.Pnom"] if "Inverter.Pnom" in table else table["Inverter.Snom"]
+      ),
+      table["Inverter.Snom"],
+      1.0
+    )
+    ```
+## Units
+Power Grid Model uses SI units (e.g. "W" for Watts),
+but source data may be supplied in different units (e.g. "MW" for Mega Watts).
+Iff units are supplied in the tabular data,
```
Comment (petersalemink95) [inline docs/converters/tabular_converter.md:107]: *typo if

File: .gitignore (hunk 9 above)
Comment (TonyXiang8787) [reply .gitignore:7]: The `.gitignore` is generated by GitHub itself. When you select a language, e.g. Python, GitHub will generate a standard `.gitignore` for common files which are ignored for Python project.

Comment (TonyXiang8787) [reply .gitignore:7]: Usually when you create a new repository from GitHub, it will generate a automatic `.gitignore` based on language selection.

Comment (bramstoeller) [reply .gitignore:7]: Not if you select .gitignore template: **None** 😛 , like I did 😉

Comment (petersalemink95) [reply .gitignore:7]: I'm not convinced of using the auto generated `.gitignore`. I think it looks messy to just have 100+ random things to ignore

Comment (TonyXiang8787) [reply .gitignore:7]: @bramstoeller @petersalemink95 it appears that both of you have similar opinion. I agree with you on this and you can delete most of the changes.

**Review by petersalemink95 — APPROVED** [review]
